drivers/adc/adc_xmc4xxx.c

```diff
@@ -0,0 +1,345 @@
+		total_buffer_size *= (1 + sequence->options->extra_samplings);
+	}
+
+	if (sequence->buffer_size < total_buffer_size) {
+		return -ENOMEM;
+	}
+
+	return 0;
+}
+
+static int start_read(const struct device *dev,
+		      const struct adc_sequence *sequence)
+{
+	int ret;
+	struct adc_xmc4xxx_data *data = dev->data;
+	const struct adc_xmc4xxx_cfg *config = dev->config;
+	XMC_VADC_GROUP_t *adc_group = config->base;
+	uint32_t requested_channels = sequence->channels;
+	uint8_t resolution = sequence->resolution;
+	uint32_t configured_channels = adc_group->ASSEL & requested_channels;
+
+	if (requested_channels == 0) {
+		LOG_ERR("No channels requested");
+		return -EINVAL;
+	}
+
+	if (requested_channels != configured_channels) {
+		LOG_ERR("Selected channels not configured");
+		return -EINVAL;
+	}
+
+	if (sequence->oversampling) {
+		LOG_ERR("Oversampling not supported");
+		return -ENOTSUP;
+	}
+
+	ret = adc_xmc4xxx_validate_buffer_size(sequence);
+	if (ret < 0) {
+		LOG_ERR("Invalid sequence buffer size");
+		return ret;
+	}
+
+	adc_group->ICLASS[0] &= ~VADC_G_ICLASS_CMS_Msk;
+	if (resolution == 8) {
+		adc_group->ICLASS[0] |=  2 << VADC_G_ICLASS_CMS_Pos;
+	} else if (resolution == 10) {
+		adc_group->ICLASS[0] |=  1 << VADC_G_ICLASS_CMS_Pos;
+	} else if (resolution == 12) {
+		adc_group->ICLASS[0] |=  0 << VADC_G_ICLASS_CMS_Pos;
+	} else {
+		LOG_ERR("Invalid resolution");
+		return -EINVAL;
+	}
+
+	data->channel_mask = requested_channels;
+	data->buffer = sequence->buffer;
+	adc_context_start_read(&data->ctx, sequence);
+
+	return adc_context_wait_for_completion(&data->ctx);
+}
+
+static int adc_xmc4xxx_read(const struct device *dev,
+			  const struct adc_sequence *sequence)
+{
+	int ret;
+	struct adc_xmc4xxx_data *data = dev->data;
+
+	adc_context_lock(&data->ctx, false, NULL);
+	ret = start_read(dev, sequence);
+	adc_context_release(&data->ctx, ret);
+	return ret;
+}
+
+#ifdef CONFIG_ADC_ASYNC
+static int adc_xmc4xxx_read_async(const struct device *dev,
+			       const struct adc_sequence *sequence,
+			       struct k_poll_signal *async)
+{
+	int ret;
+	struct adc_xmc4xxx_data *data = dev->data;
+
+	adc_context_lock(&data->ctx, true, async);
+	ret = start_read(dev, sequence);
+	adc_context_release(&data->ctx, ret);
+
+	return ret;
+}
+#endif
+
+static int adc_xmc4xxx_channel_setup(const struct device *dev,
+				   const struct adc_channel_cfg *channel_cfg)
+{
+	const struct adc_xmc4xxx_cfg *config = dev->config;
+	VADC_G_TypeDef *adc_group = config->base;
+	uint32_t ch_num = channel_cfg->channel_id;
+
+	if (ch_num >= XMC4XXX_CHANNEL_COUNT) {
+		LOG_ERR("Channel %d is not valid", ch_num);
+		return -EINVAL;
+	}
+
+	if (channel_cfg->differential) {
+		LOG_ERR("Differential channels are not supported");
+		return -EINVAL;
+	}
+
+	if (channel_cfg->gain != ADC_GAIN_1) {
+		LOG_ERR("Invalid channel gain");
+		return -EINVAL;
+	}
+
+	if (channel_cfg->reference != ADC_REF_INTERNAL) {
+		LOG_ERR("Invalid channel reference");
+		return -EINVAL;
+	}
```

Comment:
Sorry, I missed that earlier. `channel_cfg->acquisition_time` is not handled here in any way.